File: superstitioMain/java/superstitio/customStrings/StringsSetManager.java
package superstitio.customStrings;

import superstitio.DataManager;
import superstitio.SuperstitioConfig;
import superstitio.customStrings.interFace.WordReplace;
import superstitio.customStrings.stringsSet.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static superstitio.customStrings.stringsSet.CardStringsWillMakeFlavorSet.makeCardNameReplaceRules;
import static superstitio.customStrings.stringsSet.ModifierStringsSet.MakeKeyWords;
import static superstitio.customStrings.stringsSet.ModifierStringsSet.makeModifierNameReplaceRules;

public class StringsSetManager {
    private static final String[] powerFiles = new String[]{"power", "power_Lupa", "power_Maso"};
    private static final String[] modifierFiles = new String[]{"modifier_damage", "modifier_block"};
    private static final String[] cardFiles = new String[]{"card_Lupa", "card_General", "card_Maso", "card_FuckJobAndBase", "card_TempCard",
            "card_Colorless"};
    private static List<WordReplace> wordReplaceRules;

    public static void loadModifierStringsSet() {
        for (String modifierFile : modifierFiles)
            DataManager.loadCustomStringsFile(modifierFile, DataManager.modifiers, ModifierStringsSet.class);
    }

    public static void loadPowerStringsSet() {
        for (String powerFile : powerFiles)
            DataManager.loadCustomStringsFile(powerFile, DataManager.powers, PowerStringsSet.class);
    }

    public static void loadCardStringsSet() {
        for (String cardFile : cardFiles)
            DataManager.loadCustomStringsFile(cardFile, DataManager.cards, CardStringsWillMakeFlavorSet.class);
    }

    public static void loadAllStrings() {
        loadCardStringsSet();
        loadModifierStringsSet();
        loadPowerStringsSet();
        loadOrbStringsSet();
        loadUIStringsSet();

        makeSFWVersion();
    }

    public static List<WordReplace> makeWordReplaceRule() {
        if (wordReplaceRules != null && !wordReplaceRules.isEmpty()) return wordReplaceRules;
        List<WordReplace> sfwReplaces =
                Arrays.stream(DataManager.makeJsonStringFromFile("SFW_replace", WordReplace[].class)).collect(Collectors.toList());
        if (DataManager.cards != null && !DataManager.cards.isEmpty())
            sfwReplaces.addAll(makeCardNameReplaceRules(new ArrayList<>(DataManager.cards.values())));
        if (DataManager.modifiers != null && !DataManager.modifiers.isEmpty())
            sfwReplaces.addAll(makeModifierNameReplaceRules(new ArrayList<>(DataManager.modifiers.values())));
        sfwReplaces.addAll(SuperstitioKeyWord.makeKeywordNameReplaceRules(SuperstitioKeyWord.getAndRegisterKeywordsFormFile()));
        wordReplaceRules = sfwReplaces.stream().filter(wordReplace -> !wordReplace.hasNullOrEmpty()).collect(Collectors.toList());
        return wordReplaceRules;
    }

    public static void makeSFWVersion() {
        if (SuperstitioConfig.isEnableSFW()) {
            makeSFWWordForStringsSet();
        }
    }

    public static void makeKeyWords() {
        List<SuperstitioKeyWord> keywords = getAllKeywords();
        keywords.forEach(keyWord -> keyWord.makeSFWVersion(makeWordReplaceRule()));
        keywords.forEach(SuperstitioKeyWord::addToGame);
        return;
    }

    private static void loadUIStringsSet() {
        DataManager.loadCustomStringsFile("UIStringsWithSFW", DataManager.uiStrings, UIStringsSet.class);
    }

    private static void loadOrbStringsSet() {
        DataManager.loadCustomStringsFile("orb", DataManager.orbs, OrbStringsSet.class);
    }

    private static void makeSFWWordForStringsSet() {
        List<WordReplace> wordReplaces = makeWordReplaceRule();

        DataManager.cards.forEach((string, card) -> card.setupSFWStringByWordReplace(wordReplaces));
        DataManager.powers.forEach(((string, power) -> power.setupSFWStringByWordReplace(wordReplaces)));
        DataManager.modifiers.forEach(((string, modifier) -> modifier.setupSFWStringByWordReplace(wordReplaces)));
        DataManager.orbs.forEach(((string, orb) -> orb.setupSFWStringByWordReplace(wordReplaces)));

    }

    private static List<SuperstitioKeyWord> getAllKeywords() {
        List<SuperstitioKeyWord> keywordsSFW = new ArrayList<>();
        keywordsSFW.addAll(SuperstitioKeyWord.getAndRegisterKeywordsFormFile());
        keywordsSFW.addAll(Arrays.stream(MakeKeyWords()).map(SuperstitioKeyWord::STSLibKeyWordToThisType).collect(Collectors.toList()));
        return keywordsSFW;
    }
}

File: superstitioMain/java/superstitio/customStrings/StringsSetManager.kt
package superstitio.customStrings

import superstitio.DataManager
import superstitio.SuperstitioModSetup
import superstitio.customStrings.interFace.WordReplace
import superstitio.customStrings.stringsSet.*
import java.util.*

object StringsSetManager
{
    private val powerFiles = arrayOf("power", "power_Lupa", "power_Maso")
    private val modifierFiles = arrayOf("modifier_damage", "modifier_block")
    private val cardFiles = arrayOf(
        "card_Lupa", "card_General", "card_Maso", "card_FuckJobAndBase", "card_TempCard",
        "card_Colorless"
    )
    private var wordReplaceRules: MutableList<WordReplace>? = null

    @JvmStatic
    fun loadModifierStringsSet()
    {
        for (modifierFile in modifierFiles) DataManager.loadCustomStringsFile(
            modifierFile,
            DataManager.modifiers,
            ModifierStringsSet::class.java
        )
    }

    @JvmStatic
    fun loadPowerStringsSet()
    {
        for (powerFile in powerFiles) DataManager.loadCustomStringsFile(
            powerFile,
            DataManager.powers,
            PowerStringsSet::class.java
        )
    }

    @JvmStatic
    fun loadCardStringsSet()
    {
        for (cardFile in cardFiles) DataManager.loadCustomStringsFile(
            cardFile,
            DataManager.cards,
            CardStringsWillMakeFlavorSet::class.java
        )
    }

    @JvmStatic
    fun loadAllStrings()
    {
        loadCardStringsSet()
        loadModifierStringsSet()
        loadPowerStringsSet()
        loadOrbStringsSet()
        loadUIStringsSet()

        makeSFWVersion()
    }

    @JvmStatic
    fun makeWordReplaceRule(): MutableList<WordReplace>
    {
        if (SuperstitioModSetup.SEAL_MANUAL_SFW)
            return ArrayList()
        if (!wordReplaceRules.isNullOrEmpty())
            return wordReplaceRules!!
        val sfwReplaces =
            DataManager.makeJsonStringFromFile("SFW_replace", Array<WordReplace>::class.java).toMutableList()
        if (DataManager.cards.isNotEmpty())
            sfwReplaces.addAll(
                CardStringsWillMakeFlavorSet.makeCardNameReplaceRules(ArrayList(DataManager.cards.values))
            )
        if (DataManager.modifiers.isNotEmpty())
            sfwReplaces.addAll(
                ModifierStringsSet.makeModifierNameReplaceRules(ArrayList(DataManager.modifiers.values))
            )
        sfwReplaces.addAll(
            SuperstitioKeyWord.makeKeywordNameReplaceRules(SuperstitioKeyWord.getAndRegisterKeywordsFormFile())
        )
        wordReplaceRules = sfwReplaces.filterNot { it.hasNullOrEmpty() }.toMutableList()
        return wordReplaceRules!!
    }

    @JvmStatic
    fun makeSFWVersion()
    {
//        if (SuperstitioConfig.isEnableSFW()) {
        makeSFWWordForStringsSet()
        //        }
    }

    @JvmStatic
    fun makeKeyWords()
    {
        val keywords = allKeywords
        keywords.forEach { keyWord -> keyWord.makeSFWVersion(makeWordReplaceRule()) }
        keywords.forEach { obj -> obj.addToGame() }
        return
    }

    val allKeywords: List<SuperstitioKeyWord>
        get()
        {
            val keywordsSFW: MutableList<SuperstitioKeyWord> = ArrayList()
            keywordsSFW.addAll(SuperstitioKeyWord.getAndRegisterKeywordsFormFile())
            keywordsSFW.addAll(ModifierStringsSet.MakeKeyWords().map(SuperstitioKeyWord::STSLibKeyWordToThisType))
            return keywordsSFW
        }

    private fun loadUIStringsSet()
    {
        DataManager.loadCustomStringsFile(
            "UIStringsWithSFW",
            DataManager.uiStrings,
            UIStringsSet::class.java
        )
    }

    private fun loadOrbStringsSet()
    {
        DataManager.loadCustomStringsFile(
            "orb",
            DataManager.orbs,
            OrbStringsSet::class.java
        )
    }

    private fun makeSFWWordForStringsSet()
    {
        val wordReplaces = makeWordReplaceRule()

        DataManager.cards.values.forEach { it.setupSFWStringByWordReplace(wordReplaces) }
        DataManager.powers.values.forEach { it.setupSFWStringByWordReplace(wordReplaces) }
        DataManager.modifiers.values.forEach { it.setupSFWStringByWordReplace(wordReplaces) }
        DataManager.orbs.values.forEach { it.setupSFWStringByWordReplace(wordReplaces) }
    }
}
